Source/Core/VideoCommon/OpcodeDecoding.cpp

```diff
@@ -228,6 +228,24 @@ class RunCallback final : public Callback
       return g_main_cp_state;
   }
 
+  OPCODE_CALLBACK(u32 GetVertexSize(u8 vat))
+  {
+    const BitSet8& attr_dirty = is_preprocess ? VertexLoaderManager::g_preprocess_vat_dirty :
+                                                VertexLoaderManager::g_main_vat_dirty;
+    if (!attr_dirty[vat]) [[likely]]
+    {
+      const auto& vertex_loaders = is_preprocess ?
+                                       VertexLoaderManager::g_preprocess_vertex_loaders :
+                                       VertexLoaderManager::g_main_vertex_loaders;
+      const VertexLoaderBase* loader = vertex_loaders[vat];
+      return loader->m_vertex_size;
+    }
+    else
+    {
+      return VertexLoaderBase::GetVertexSize(GetCPState().vtx_desc, GetCPState().vtx_attr[vat]);
+    }
+  }
```

Comment:
IMO it would be better to use `VertexLoaderManager::RefreshLoader<is_preprocess>(vat)->m_vertex_size`. It looks like you were partially working on doing that (since you added `RefreshLoader` to the header), but didn't actually end up doing it.

If `GetVertexSize` is being called, that means that `OnPrimitiveCommand` will be called as soon as there is enough bytes for all of the vertices. `OnPrimitiveCommand` calls `VertexLoaderManager::RunVertices`, which will also call `RefreshLoader`, and in that case `attr_dirty` would no longer be true. This would mean that in general new vertex loaders would be JITed in `GetVertexSize` instead of `RunVertices`, but that doesn't matter too much.

(Theoretically the `RefreshLoader` call in `RunVertices` would no longer even be needed, as it would be known that the appropriate loader already exists. But I don't know if there's a good way of doing that without making things more jank.)

Reply:
Done. I needed to move some stuff around to make it as fast as the other solution.